// @ngneat/error-tailor/esm2020/lib/control-error-anchor.directive.mjs
import { Directive, inject, ViewContainerRef } from '@angular/core';
import * as i0 from "@angular/core";
export class ControlErrorAnchorDirective {
    constructor() {
        this.vcr = inject(ViewContainerRef);
    }
}
ControlErrorAnchorDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: ControlErrorAnchorDirective, deps: [], target: i0.ɵɵFactoryTarget.Directive });
ControlErrorAnchorDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "15.0.1", type: ControlErrorAnchorDirective, isStandalone: true, selector: "[controlErrorAnchor]", exportAs: ["controlErrorAnchor"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: ControlErrorAnchorDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[controlErrorAnchor]',
                    standalone: true,
                    exportAs: 'controlErrorAnchor'
                }]
        }] });
