// @ngneat/error-tailor/esm2020/lib/error-tailor.providers.mjs
import { NgModule } from '@angular/core';
import { ControlErrorAnchorDirective } from './control-error-anchor.directive';
import { DefaultControlErrorComponent } from './control-error.component';
import { ControlErrorsDirective } from './control-error.directive';
import { FormActionDirective } from './form-action.directive';
import { ErrorTailorConfigProvider, FORM_ERRORS } from './providers';
import * as i0 from "@angular/core";
const _errorTailorImports = [
    ControlErrorsDirective,
    ControlErrorAnchorDirective,
    DefaultControlErrorComponent,
    FormActionDirective
];
export class errorTailorImports {
}
errorTailorImports.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: errorTailorImports, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
errorTailorImports.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.0.1", ngImport: i0, type: errorTailorImports, imports: [ControlErrorsDirective,
        ControlErrorAnchorDirective,
        DefaultControlErrorComponent,
        FormActionDirective], exports: [ControlErrorsDirective,
        ControlErrorAnchorDirective,
        DefaultControlErrorComponent,
        FormActionDirective] });
errorTailorImports.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: errorTailorImports, imports: [DefaultControlErrorComponent] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.0.1", ngImport: i0, type: errorTailorImports, decorators: [{
            type: NgModule,
            args: [{
                    imports: [_errorTailorImports],
                    exports: [_errorTailorImports]
                }]
        }] });
export function provideErrorTailorConfig(config) {
    return [
        {
            provide: ErrorTailorConfigProvider,
            useValue: config
        },
        {
            provide: FORM_ERRORS,
            ...config.errors
        }
    ];
}
